File: src/main/java/club/libridge/libridgebackend/core/Strain.java
package club.libridge.libridgebackend.core;

import java.util.EnumMap;
import java.util.HashMap;
import java.util.Map;

import club.libridge.libridgebackend.core.rulesets.concrete.PositiveNoTrumpsRuleset;
import club.libridge.libridgebackend.core.rulesets.concrete.PositiveRuleset;
import club.libridge.libridgebackend.core.rulesets.concrete.PositiveWithTrumpsRuleset;
import lombok.AllArgsConstructor;
import lombok.Getter;
import lombok.NonNull;

/**
 * The Laws of Bridge use "Denomination" but we will use Strain as it is more
 * common.
 */
@AllArgsConstructor
@Getter
public enum Strain {
    CLUBS("Clubs", new PositiveWithTrumpsRuleset(Suit.CLUBS)), DIAMONDS("Diamonds", new PositiveWithTrumpsRuleset(Suit.DIAMONDS)),
    HEARTS("Hearts", new PositiveWithTrumpsRuleset(Suit.HEARTS)), SPADES("Spades", new PositiveWithTrumpsRuleset(Suit.SPADES)),
    NOTRUMPS("No Trumps", new PositiveNoTrumpsRuleset());

    private final String name;
    private final PositiveRuleset positiveRuleset;

    public String getSymbol() {
        return String.valueOf(this.name.charAt(0));
    }

    private static final Map<String, Strain> MAP_FROM_NAME;
    private static final Map<Suit, Strain> MAP_FROM_SUIT;

    static {
        MAP_FROM_NAME = new HashMap<String, Strain>();
        for (Strain strain : Strain.values()) {
            MAP_FROM_NAME.put(strain.getName(), strain);
        }

        MAP_FROM_SUIT = new EnumMap<Suit, Strain>(Suit.class);
        MAP_FROM_SUIT.put(Suit.CLUBS, CLUBS);
        MAP_FROM_SUIT.put(Suit.DIAMONDS, DIAMONDS);
        MAP_FROM_SUIT.put(Suit.HEARTS, HEARTS);
        MAP_FROM_SUIT.put(Suit.SPADES, SPADES);
    }

    public static Strain fromName(@NonNull String name) {
        return Strain.MAP_FROM_NAME.get(name);
    }

    public static Strain fromSuit(@NonNull Suit suit) {
        return Strain.MAP_FROM_SUIT.get(suit);
    }

}

File: src/main/scala/scalabridge/Strain.scala
package scalabridge

import scalabridge.exceptions.StrainException
import scala.collection.immutable.ListMap
import scala.util.Try
import scala.util.Success
import scala.util.Failure

/** 
  * The Laws of Bridge use "Denomination" but we will use Strain as it is more
  * common.
  * LAW 18 - BIDS specify the order of the Strains. We use it here.
  */
enum Strain(name: String) extends java.lang.Enum[Strain] {
  case CLUBS extends Strain("Clubs")
  case DIAMONDS extends Strain("Diamonds")
  case HEARTS extends Strain("Hearts")
  case SPADES extends Strain("Spades")
  case NOTRUMPS extends Strain("No Trumps")
  def getName: String = this.name
  def getSymbol: String = String.valueOf(this.name.charAt(0))
}
object Strain {
  def getCLUBS() = CLUBS
  def getDIAMONDS() = DIAMONDS
  def getHEARTS() = HEARTS
  def getSPADES() = SPADES
  def getNOTRUMPS() = NOTRUMPS
  def fromSuit(suit: Suit): Strain =
    suit match
      case scalabridge.Suit.CLUBS    => Strain.CLUBS
      case scalabridge.Suit.DIAMONDS => Strain.DIAMONDS
      case scalabridge.Suit.HEARTS   => Strain.HEARTS
      case scalabridge.Suit.SPADES   => Strain.SPADES

  // Guessing a ListMap is faster than a HashMap here. If this becomes an issue:
  // Benchmark it, write the results here and refactor.
  // Static copy to avoid many copies
  private val nameToStrainMap = ListMap.empty ++ Strain.values
    .map(strain => strain.getSymbol.substring(0, 1) -> strain)
    .toMap

  def fromName(name: String): Try[Strain] = {
    nameToStrainMap.get(name.toUpperCase()) match
      case Some(value) => Success(value)
      case None        => Failure(StrainException(name))
  }
}
